### PR TITLE
feat: feature-flagged cross-chain swaps route [METABRIDGE-867]

### PR DESCRIPTION
**Pull request by micaelae**
## **Description**
This change adds a placeholder route for the cross-chain swaps experience (roughly based on swaps page). Clicking on the Bridge button should link to the route when the feature flag is turned on. Otherwise, the portfolio bridge experience is loaded (current behavior)

New e2e tests: https://github.com/MetaMask/metamask-extension/pull/25812

Migration to BaseController v2: https://github.com/MetaMask/metamask-extension/pull/26109/

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25481?quickstart=1)

## **Related issues**

Fixes: [METABRIDGE-867](https://consensyssoftware.atlassian.net/browse/METABRIDGE-867)

## **Manual testing steps**

1. Build locally and set `BRIDGE_USE_DEV_APIS=1`
2. Check that bridge-api getAllFeatureFlags is fetched
3. Verify that bridgeState FF  is set to true in state logs
3. Click on Bridge button and verify Bridge page is visible
5. Build locally and set `BRIDGE_USE_DEV_APIS=0`
6. Wait ~10 minutes
7. Reload and Check that bridge-api getAllFeatureFlags is fetched
3. Verify that bridgeState FF  is set to false in state logs
8. Click on Bridge button and verify Portfolio URL is opened in a new tab

### **Before**
N/A

### **After**

Screenshot 2024-07-12 at 3 44 34 PM

## **Pre-merge author checklist**

- [X] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
 Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

[METABRIDGE-867]: https://consensyssoftware.atlassian.net/browse/METABRIDGE-867?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: ui/pages/asset/components/token-buttons.tsx [hunk 1]
```diff
@@ -31,6 +31,10 @@ import {
   getMetaMetricsId,
   ///: END:ONLY_INCLUDE_IF
 } from '../../../selectors';
+///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)
```
Comment (BZahory) [inline ui/pages/asset/components/token-buttons.tsx:32]: What's the reason for the code fence here?

File: ui/pages/asset/components/token-buttons.tsx [hunk 2]
```diff
@@ -74,6 +78,7 @@ const TokenButtons = ({
   const isBuyableChain = useSelector(getIsNativeTokenBuyable);
   const metaMetricsId = useSelector(getMetaMetricsId);
   const { openBuyCryptoInPdapp } = useRamps();
+  useBridging();
```
Comment (BZahory) [inline ui/pages/asset/components/token-buttons.tsx]: On a similar note, I think we would need to code fence both or neither. Or we might call `useBridging` without importing it

File: ui/pages/bridge/index.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      <div className="bridge__container">
+        <div className="bridge__header">
+          <Box
+            display={Display.Flex}
+            justifyContent={JustifyContent.center}
+            marginLeft={4}
+            width={BlockSize.OneTwelfth}
+            tabIndex={0}
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            onKeyUp={(e: any) => {
```
Comment (BZahory) [inline ui/pages/bridge/index.tsx:61]: You can type this as `React.KeyboardEvent<HTMLInputElement>`

File: ui/pages/bridge/index.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  const history = useHistory();
+  const dispatch = useDispatch();
+
+  const isBridgeEnabled = useSelector(getIsBridgeEnabled);
+
+  const redirectToDefaultRoute = async () => {
+    history.push({
+      pathname: DEFAULT_ROUTE,
+      // @ts-expect-error - property 'state' does not exist on type PartialPath.
+      state: { stayOnHomePage: true },
```
Comment (BZahory) [inline ui/pages/bridge/index.tsx:44]: I noticed the second `history.push` parameter is `state`. Does this accomplish the same thing being done here?

File: ui/components/app/wallet-overview/coin-buttons.tsx [hunk 5]
```diff
@@ -284,29 +284,20 @@ const CoinButtons = ({
   }, [chainId, defaultSwapsToken]);
 
   const handleBridgeOnClick = useCallback(() => {
-    if (isBridgeChain) {
-      const portfolioUrl = getPortfolioUrl(
-        'bridge',
-        'ext_bridge_button',
-        metaMetricsId,
+    defaultSwapsToken &&
```
Comment (BZahory) [inline ui/components/app/wallet-overview/coin-buttons.tsx:287]: Could we use an early return here instead?

File: ui/pages/bridge/index.tsx (hunk 3 above)
Comment (micaelae) [reply ui/pages/bridge/index.tsx:61]: nice!

File: ui/pages/bridge/index.tsx (hunk 4 above)
Comment (micaelae) [reply ui/pages/bridge/index.tsx:44]: The second parameter prevents the DEFAULT_ROUTE (home) from redirecting to other pages based on any pending txs, notifications, etc

File: privacy-snapshot.json [hunk 6]
```diff
@@ -7,6 +7,7 @@
   "bafkreifvhjdf6ve4jfv6qytqtux5nd4nwnelioeiqx5x2ez5yrgrzk7ypi.ipfs.dweb.link",
   "bafybeidxfmwycgzcp4v2togflpqh2gnibuexjy4m4qqwxp7nh3jx5zlh4y.ipfs.dweb.link",
   "bridge.api.cx.metamask.io",
+  "bridge.dev-api.cx.metamask.io",
```
[the remark this answers is not in the document]
Comment (micaelae) [reply privacy-snapshot.json:10]: good call, this can be removed

File: ui/pages/bridge/index.tsx (hunk 4 above)
Comment (BZahory) [reply ui/pages/bridge/index.tsx:44]: I was referring to moving the `state` option from the first argument into the second argument, so we could avoid the `ts-expect-error`

**Review by BZahory — DISMISSED** [review]

**Review by BZahory — DISMISSED** [review]

**Review by BZahory — APPROVED** [review]

File: test/e2e/tests/bridge/bridge-test-utils.ts [hunk 7]
```diff
@@ -120,7 +122,7 @@ export const getBridgeFixtures = (
 
   return {
     driverOptions: {
-      openDevToolsForTabs: true,
+      // openDevToolsForTabs: true,
```
Comment (pedronfigueiredo) [inline test/e2e/tests/bridge/bridge-test-utils.ts:125]: Did you intend to delete this line?

**Review by pedronfigueiredo — APPROVED** [review]
LGTM

File: test/e2e/tests/bridge/bridge-test-utils.ts (hunk 7 above)
Comment (micaelae) [reply test/e2e/tests/bridge/bridge-test-utils.ts:125]: Thanks for taking a look! This is useful for local debugging, but it makes the tests flaky in CI so I commented it out

**Pull request merged**